File: geekbang/algorithm/source-code/3/lc105 从前序与中序遍历序列构造二叉树.go
	// 前序遍历中的第一个节点就是根节点
	// 在中序遍历中定位根节点，然后确认左右子树的范围
	// 递归地构造左子树，并连接到根节点
	// 递归地构造右子树，并连接到根节点
	func buildTree(preorder []int, inorder []int) *TreeNode {
		// 递归终止条件
		if len(preorder) == 0 {
			return nil
		}
		// 当前层的处理逻辑，从先序数组中确认根节点
		root := &TreeNode{preorder[0], nil, nil}
		i := 0
		// 找到中序中根节点的位置
		for ; i < len(inorder); i++ {
			if inorder[i] == preorder[0] {
				break
			}
		}
		// 递归创建左子树
		root.Left = buildTree(preorder[1:len(inorder[:i])+1], inorder[:i])
		// 递归创建右子树
		root.Right = buildTree(preorder[len(inorder[:i])+1:], inorder[i+1:])
		return root
	}

File: geekbang/algorithm/source-code/3/lc105 从前序与中序遍历序列构造二叉树.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def buildTree(self, preorder: List[int], inorder: List[int]) -> TreeNode:
        self.preorder = preorder
        self.inorder = inorder
        return self.build(0, len(preorder) - 1, 0, len(inorder) - 1)

    # 用 preorder[l1..r1] 和 inorder[l2..r2] 还原二叉树
    def build(self, l1, r1, l2, r2):
        if l1 > r1:
            return None
        root = TreeNode(self.preorder[l1])
        mid = l2 # mid是root在inorder中的位置
        while self.inorder[mid] != root.val:
            mid += 1
        # l2~mid-1就是左子树中序
        # mid+1~r2就是右子树中序
        root.left = self.build(l1 + 1, l1 + (mid - l2), l2, mid - 1)
        root.right = self.build(l1 + (mid - l2) + 1, r1, mid + 1, r2)
        return root